### PR TITLE
RavenDB-9919 - Problem migration indexes from 3.5 automatically

### PR DESCRIPTION
**Pull request by grisha-kotler**
http://issues.hibernatingrhinos.com/issue/RavenDB-9919

### CODE REVIEW

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs [hunk 1]
```diff
@@ -351,6 +338,73 @@ private SmugglerProgressBase.Counts ProcessIndexes(SmugglerResult result)
...
+
+                actions.WriteIndex(indexDefinition);
+            }
+            catch (Exception e)
+            {
+                var exceptionMessage = e.Message;
+                if (exceptionMessage.Contains("No overload for method 'LoadDocument' takes 1 arguments"))
+                {
+                    exceptionMessage = $"{e.Message}" + Environment.NewLine +
+                            "LoadDocument requires a generic argument for its usage (the type of the loaded document)" + Environment.NewLine +
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs:359]: Shouldn't this be: LoadDocument requires a second argument which is a collection name of the loaded document (or something like that) ?

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs [hunk 2]
```diff
@@ -351,6 +338,73 @@ private SmugglerProgressBase.Counts ProcessIndexes(SmugglerResult result)
...
+                {
+                    errorMessage += Environment.NewLine + $"Reduce: {indexDefinition.Reduce}";
+                }
+                result.AddError(errorMessage);
+            }
+
+            string ReplaceLegacyMethods(string str)
+            {
+                return str
+                    .Replace("AbstractIndexCreationTask.SpatialGenerate", "CreateSpatialField")
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs:402]: Let's not do that, this will not work as expected, the arguments do not match.

**Review by ppekrol — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs [hunk 3]
```diff
@@ -351,6 +338,73 @@ private SmugglerProgressBase.Counts ProcessIndexes(SmugglerResult result)
...
+                                "\tselect new {" + Environment.NewLine +
+                                    "\t\tCompanyName: comapny.Name" + Environment.NewLine +
+                                "\t}" + Environment.NewLine;
+                }
+                else if (retryOnError && 
+                         (exceptionMessage.Contains("The name 'AbstractIndexCreationTask' does not exist in the current context") ||
+                          exceptionMessage.Contains("The name 'SpatialIndex' does not exist in the current context") ||
+                          exceptionMessage.Contains("The type or namespace name 'Abstractions' does not exist in the namespace 'Raven' (are you missing an assembly reference?)")))
+                {
+                    var maps = new HashSet<string>();
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs:372]: I do not think we should try to fix the index on-fly.

Comment (grisha-kotler) [reply src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs:372]: we can save the index as faulty index and allow the user manually edit it.

Comment (ppekrol) [reply src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs:372]: I do not think we want to do that, this is not a recommended way of migrating indexes.

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs:402]: for those that don't match we display an error:
(22,74): error CS1501: No overload for method 'CreateSpatialField' takes 4 arguments

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs (hunk 3 above)
Comment (ayende) [reply src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs:372]: Yes, better to error clearly

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs:359]: done

**Pull request closed without merging**